swarm/storage/mru/resource.go

```diff
@@ -318,87 +321,119 @@ func (h *Handler) chunkSize() int64 {
+			rootAddr: chunk.Addr,
+		},
+		resourceMetadata: resourceMetadata{
+			name:      request.name,
+			startTime: request.startTime,
+			frequency: request.frequency,
+		},
+		updated: time.Now(),
+	}
+	copy(rsrc.ownerAddr[:], request.ownerAddr[:])
+	h.set(chunk.Addr, rsrc)
+
+	return nil
+}
+
+func (h *Handler) NewUpdateRequest(ctx context.Context, rootAddr storage.Address) (*UpdateRequest, error) {
+
+	if rootAddr == nil {
+		return nil, NewError(ErrInvalidValue, "rootAddr cannot be nil")
 	}
-	copy(rsrc.owner[:], ownerAddr[:])
-	h.set(chunk.Addr.Hex(), rsrc)
 
-	return chunk.Addr, rsrc, nil
+	rsrc, err := h.Load(rootAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	currentblock := h.getCurrentTime(ctx)
+
+	mru := new(UpdateRequest)
+	mru.period, err = getNextPeriod(rsrc.startTime, currentblock, rsrc.frequency)
+	if err != nil {
+		return nil, err
+	}
+	if _, err = h.lookup(rsrc, mru.period, 0, 0); err != nil {
+		return nil, err
+	}
+
+	if !rsrc.isSynced() {
+		return nil, NewError(ErrNotSynced, fmt.Sprintf("NewMruRequest: object '%s' not in sync", rootAddr.Hex()))
+	}
+
+	mru.name = rsrc.name
+	mru.multihash = rsrc.multihash
+	mru.rootAddr = rsrc.rootAddr
+	mru.metaHash = rsrc.metaHash
+	mru.resourceMetadata = rsrc.resourceMetadata
+
+	// if we already have an update for this block then increment version
+	// resource object MUST be in sync for version to be correct, but we checked this earlier in the method already
+	if h.hasUpdate(rootAddr, mru.period) {
+		mru.version = rsrc.version + 1
+	} else {
+		mru.version = 1
+	}
+
+	return mru, nil
 }
 
 // creates a metadata chunk
-func (h *Handler) newMetaChunk(name string, startBlock uint64, frequency uint64, ownerAddr common.Address) *storage.Chunk {
+func (h *Handler) newMetaChunk(metadata *resourceMetadata) (chunk *storage.Chunk, metaHash []byte) {
 	// the metadata chunk points to data of first blockheight + update frequency
 	// from this we know from what blockheight we should look for updates, and how often
 	// it also contains the name of the resource, so we know what resource we are working with
-	metadataChunkLength := metadataChunkOffsetSize + len(name)
-	data := make([]byte, metadataChunkLength)
-
-	// root block has first two bytes both set to 0, which distinguishes from update bytes
-	val := make([]byte, 8)
-	binary.LittleEndian.PutUint64(val, startBlock)
-	copy(data[:8], val)
-	binary.LittleEndian.PutUint64(val, frequency)
-	copy(data[8:16], val)
-	copy(data[16:], ownerAddr[:])
-	copy(data[16+common.AddressLength:], []byte(name))
-	binary.LittleEndian.PutUint64(val, uint64(metadataChunkLength))
 
 	// the key of the metadata chunk is content-addressed
 	// if it wasn't we couldn't replace it later
 	// resolving this relationship is left up to external agents (for example ENS)
-	hasher := h.hashPool.Get().(storage.SwarmHash)
-	hasher.ResetWithLength(val)
-	hasher.Write(data)
-	key := hasher.Sum(nil)
-	h.hashPool.Put(hasher)
+	rootAddr, metaHash, chunkData := metadata.hash()
```

Comment:
Moved all this to a new method of the new `resourceMetadata` structure. This method hashes and serializes the metadata in one go.
In later steps, this is further refactored, and `newMetaChunk` becomes a direct method of `resourceMetadata`